### ISSUE TITLE
[Bug]: Main loop integration with IPyhton broken after matplotlib version 3.6.2

### ISSUE THREAD

**Issue by dnicolodi** [matplotlib/matplotlib#25485]
### Bug summary

After matplotlib release 3.6.2 (I tested releases 3.6.2, 3.7.0, and 3.7.1) the main loop integration with IPython does not work as expected. I use the the `%matplotlib` IPython magic to install the main loop integration and the `%run` magic to run a script that creates some figures and calls `matplotlib.pyplot.show()`.

### Code for reproduction

```python
import matplotlib.pyplot as plt
plt.figure()
plt.show()
```

### Actual outcome

The figures are shown and the script execution does not terminate till all the figures are closed. Once all the figures created in the script are closed, the execution returns to the IPython prompt and an additional empty figure is displayed.

### Expected outcome

The script execution should terminate and the figures should be available for interaction. No extra figure should be created. This is the behavior with matplotlib 3.6.2. Conda does not have a package for matplotlib 3.6.3, thus I haven't tested with that version.

### Operating system

Windows

### Matplotlib Version

3.70, 3.71

### Matplotlib Backend

QtAgg

### Python version

3.10

### Jupyter version

Testd with IPython 8.0 to 8.10

### Installation

conda

**Comment by tacaswell** [comment]
If you add `plt.ion()` does it work again?

https://matplotlib.org/stable/users/explain/interactive.html#ipython-integration for lots of details in interactive mode.


Can you please include enough code to reproduce the problem you describe?

**Comment by dnicolodi** [comment]
The code in the original report is enough to reproduce the problem. Just drop it into a `test.py` file and call it with `%run test.py` from the IPython console prompt. I think I know how interactive mode works. What I am reporting is a change in behavior between different versions of matplotlib. Anyhow, even assuming that the fact that the script does not terminate execution leaving the handling of the redisplay loop to IPython is an issue with how the interactive mode is setup, having an empty figure pop up is definitely a bug somewhere.

**Comment by dnicolodi** [comment]
I can reproduce the same issue on macOS with the MacOSX backend. Calling `plt.ion()` from the IPython prompt and then running the script with `%run` does not change the behavior. Adding `plt.ion()` to the script just before `plt.show()` fixes the issue. However, having `plt.ion()` in the script was not necessary with matplotlib 3.6, and having the script work without having to add it IMHO is the more correct behavior.

**Comment by tacaswell** [comment]
I can not reproduce the extra window and calling `plt.ion()` does change the behavior for me with mpl with 3.6.3 (from arch packages) with the qt backend nor with 3.7.1 from conda with the OSX backend.

**Comment by dnicolodi** [comment]
I'm not sure I understand what you mean (there is a "nor" that does not match the "does change" verb coming before). Do you mean that the script execution terminates with the figures still displayed, for both matplolib 3.6.3 and 3.7.1, and that `plt.ion()` does not do change anything using either matplotlib 3.6.3 and 3.7.1?

**Comment by tacaswell** [comment]
drat, sorry doing too many things in parallel. I wrote that message in 3 chunks and finished a different sentence than I remembered starting 🤦🏻 

- I can not reproduce the extra window issue on either linux + mpl3.6.3 + qt or mac + mpl3.7.1 + osx backend
- calling `plt.ion()` before calling `%run` made `plt.show()` non-blocking in both of my test cases

In both cases this is the behavior I expect. If you do not opt-into interactive mode either manually or my setting it in your rcparams) `plt.show()` will block (becuase if you run `python test.py` then if `plt.show()` does not block the script will exit without showing you the figures which is very rarely the behavior people want.  If you do opt into interactive mode (which makes `plt.show()` non-blocking) then you get the prompt back immediately.

**Comment by dnicolodi** [comment]
The code that causes the issue is this code in IPython that is used to run the script when `%run` is used after calling `%matplotlib` https://github.com/ipython/ipython/blob/e35dbb4e3828d643d7ffd58747350d7c90134706/IPython/core/pylabtools.py#L198-L226 which reads
```python
    def mpl_execfile(fname,*where,**kw):
        """matplotlib-aware wrapper around safe_execfile.
        Its interface is identical to that of the :func:`execfile` builtin.
        This is ultimately a call to execfile(), but wrapped in safeties to
        properly handle interactive rendering."""

        import matplotlib
        import matplotlib.pyplot as plt

        #print '*** Matplotlib runner ***' # dbg
        # turn off rendering until end of script
        is_interactive = matplotlib.rcParams['interactive']
        matplotlib.interactive(False)
        safe_execfile(fname,*where,**kw)
        matplotlib.interactive(is_interactive)
        # make rendering call now, if the user tried to do it
        if plt.draw_if_interactive.called:
            plt.draw()
            plt.draw_if_interactive.called = False

        # re-draw everything that is stale
        try:
            da = plt.draw_all
        except AttributeError:
            pass
        else:
            da()
```
So switching off interactive mode before running some code and then switching it back on works on matplotlib 3.6 but not on matplotlib 3.7, at least on my systems. The call to `plt.draw()` is what makes the extra figure pop up.

**Comment by dnicolodi** [comment]
> If you do not opt-into interactive mode either manually or my setting it in your rcparams) `plt.show()` will block

Agreed. I implicitly enable interactive mode with the `%matplotlib` magic, which is responsible for installing the execution hook above and for setting up interactive mode, among other things probably. Can you please try to reproduce calling `%matplotlib` before `%run`?

**Comment by dnicolodi** [comment]
After all, this looks like a bug in IPython caused by some changes in the matplotlib internals. Reported here https://github.com/ipython/ipython/issues/13977

**Comment by tacaswell** [comment]
After a bit of digging, I suspect that this is a collaborative bug
In pyplot we do https://github.com/matplotlib/matplotlib/blob/78bf53caacbb5ce0dc7aa73f07a74c99f1ed919b/lib/matplotlib/backend_bases.py#L2929-L2935 so either we are failing to run that logic when we should or we are somehow undoing the `_needmain` patching that IPython is doing.


-----

To clarify 3..6.x works and 3.7.x does not?  I think I miss-read the initial post as saying 3.6.2 is the first broken version (not the last working version) and hence was very skeptical that we broke this in a patch release (that is also why I tested with 3.6.3 on linux).

**Comment by tacaswell** [comment]
90% the motivation for the two sides of this are:

 - in `%run` IPython wants `draw_if_interactive` to be a no-op (which is why @dnicolodi reports the different run times)
 - but in `%run` we want the default Do What I Mean behavior of `plt.show()` to be non-blocking

**Comment by dnicolodi** [comment]
> To clarify 3..6.x works and 3.7.x does not?

Correct.

**Comment by tacaswell** [comment]
Sorry for being skeptical.

**Comment by dnicolodi** [comment]
> so either we are failing to run that logic

I've tried to debug this further, and indeed that code is not executed when the script calls `plt.plot()`.
